src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -4774,11 +4781,15 @@ public Type visitCapturedType(CapturedType t, Void s) {
 
         @Override
         public Type visitTypeVar(TypeVar t, Void s) {
-            if (rewriteTypeVars) {
-                Type bound = t.getUpperBound().contains(t) ?
-                        erasure(t.getUpperBound()) :
-                        visit(t.getUpperBound());
-                return rewriteAsWildcardType(bound, t, EXTENDS);
+            if (seen.add(t)) {
```

Comment:
IIRC, this code is called by `Types.isCastable`, in an attempt to do some fancy logic (outside the spec) which allow to determine whether two types are disjoint or not. To do this, the code attempts to rewrite occurrences of type-variables using wildcards. So, in some cases the logic fails with SOE - so that result should lead to the consequence that we don't know whether the types are disjoint or not, but, since we're outside the scope of the spec, we need to treat them as _not_ disjoint (e.g. the cast should be possible and we gave up proving that it should NOT be possible).

So, I'm afraid that `return t` here is going to make this method stricter than intended, which will lead javac to think that the types are in fact disjoint, and that will result in a compiler error. I wonder if we should we return a `?` instead?